File: ml-benchmark/autolifter/single-pass/longest1s.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;
Config SampleSize = 20;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

longest1s = fix (
  \f: Int -> List -> Int. \len: Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let len = if == h 1 then + len 1 else 0 in
    max len (f len t)
  end
) 0;

main = single_pass longest1s;

File: ml-benchmark/autolifter/single-pass/longest1s.ml
type list = Cons of int * list | Nil of unit

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) ->
      Cons (h, run t)
  in
  fun xs ->
    v (run xs)

let inf = 100

let max a b =
  if a < b then b else a

let longest1s =
  let rec f len xs =
    match xs with
    | Nil _ -> 0
    | Cons (h, t) ->
      let len = if h = 1 then len + 1 else 0 in
      max len (f len t)
  in
  fun xs -> f 0 xs

let program = single_pass longest1s